### PR TITLE
sys/hashes: add AES CMAC implementation

### PR DESCRIPTION
**Pull request by jia200x**
Hi.

This is an AES CMAC implementation with tests ([RFC4493]).
It's required for LoRaMAC.

Cheers!
[RFC4493]:https://tools.ietf.org/html/rfc4493

### CODE REVIEW

File: sys/hashes/cmac.c [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+ * @file
+ * @brief       AES_CMAC implementation
+ *
+ * @author      José Ignacio Alamos <jose.alamos@inria.cl>
+ *
+ * @}
+ */
+
+
+#include <stdio.h>
```
Comment (kaspar030) [inline sys/hashes/cmac.c]: please sort the includes (standard headers first)

File: sys/hashes/cmac.c [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+#include "hashes/cmac.h"
+#include <inttypes.h>
+#include "crypto/ciphers.h"
+#include <string.h>
+
+#define XOR128(x, y) do { \
```
Comment (kaspar030) [inline sys/hashes/cmac.c]: Could you try to make this a static inline function and see if it would increase codesize?

File: sys/hashes/cmac.c [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+        uint8_t i; \
+        for (i = 0; i < 16; i++) { \
+            (y)[i] = x[i] ^ y[i]; \
+        } \
+} while (0)
+
+#define MIN(a, b) a < b ? a : b
+
+#define LEFTSHIFT(x, y) do { \
```
Comment (kaspar030) [inline sys/hashes/cmac.c]: same here

File: sys/include/hashes/cmac.h [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+ * @file
+ * @brief       AES_CMAC interface definition
+ *
+ * @author      José Ignacio Alamos <jose.alamos@inria.cl>
+ */
+
+//TODO: Add source
+
+#include "crypto/ciphers.h"
+#include <stdio.h>
```
Comment (kaspar030) [inline sys/include/hashes/cmac.h]: please add space here

Comment (kaspar030) [inline sys/include/hashes/cmac.h]: and order the includes

File: sys/include/hashes/cmac.h [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+#define CMAC_BLOCK_SIZE 16
```
Comment (kaspar030) [inline sys/include/hashes/cmac.h]: another empty line after this

File: sys/include/hashes/cmac.h [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+typedef struct cmac_context_t {
```
Comment (kaspar030) [inline sys/include/hashes/cmac.h]: drop ```cmac_context_t``` here

File: tests/unittests/tests-hashes/tests-hashes-cmac.c [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+ *
+ * @file
+ * @brief       Test cases for the AES-CMAC hash implementation
+ *
+ * @author      José Ignacio Alamos <jose.alamos@inria.cl>
+ *
+ * @}
+ */
+
+#include "tests-hashes.h"
```
Comment (kaspar030) [inline tests/unittests/tests-hashes/tests-hashes-cmac.c]: sort includes

File: sys/hashes/cmac.c (hunk 2 above)
Comment (jia200x) [inline sys/hashes/cmac.c]: ok! I will check :P

Comment (jia200x) [inline sys/hashes/cmac.c]: I just made it static. It reduces the codesize.

**Comment by jia200x** [comment]
I will squash in case everything is OK ;)

File: sys/hashes/cmac.c [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+#include <stdio.h>
+#include <inttypes.h>
+#include <string.h>
+
+#include "hashes/cmac.h"
+#include "crypto/ciphers.h"
+
+#define MIN(a, b) a < b ? a : b
+
+void _xor128(uint8_t *x, uint8_t *y)
```
Comment (BytesGalore) [inline sys/hashes/cmac.c]: Please make this internally used helper function and `_leftshift(...)` `static`.
I'm curious, did `inline` the functions impact the codesize?

Comment (BytesGalore) [inline sys/hashes/cmac.c]: Oh, I think you missed to add the `inline` keyword in https://github.com/RIOT-OS/RIOT/pull/6109/commits/c229a517d546c5f11e2ee6c62f7f7b348897194f, this way the defines are just substituted by usual function calls.

Comment (jia200x) [inline sys/hashes/cmac.c]: Oh, I forgot to add them.
WIP

Comment (jia200x) [inline sys/hashes/cmac.c]: Done. There aren't changes in codesize between void and static inline void in this case.
But there are with the old macros.

Cheers!

**Comment by jia200x** [comment]
rebased.

**Comment by miri64** [comment]
It's clear to me now from reading the code, but would have be great if the title and description would hint somewhere that this is a hashing algorithm (and not a MAC layer; what I first thought).

File: sys/hashes/cmac.c (hunk 1 above)
Comment (aabadie) [inline sys/hashes/cmac.c]: Can you address this comment @jia200x ?

File: sys/hashes/cmac.c [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+
+        if (ctx->M_n < CMAC_BLOCK_SIZE) {
+            break;
+        }
+    }
+}
+
+void cmac_final(cmac_context_t *ctx, void *digest)
+{
+    //Generate subkeys
```
Comment (aabadie) [inline sys/hashes/cmac.c]: Please use C style comments

File: sys/hashes/cmac.c [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+    if (L[0] & 0x80) {
+        _leftshift(L, K);
+        K[15] ^= 0x87;
+    }
+    else {
+        _leftshift(L, K);
+    }
+
+    if (ctx->M_n != 16) {
+        //Generate K2
```
Comment (aabadie) [inline sys/hashes/cmac.c]: same here

File: sys/hashes/cmac.c [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+        if (K[0] & 0x80) {
+            _leftshift(K, K);
+            K[15] ^= 0x87;
+        }
+        else {
+            _leftshift(K, K);
+        }
+        //Padding
```
Comment (aabadie) [inline sys/hashes/cmac.c]: and here

File: sys/include/hashes/cmac.h [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+
+ * @{
+ *
+ * @file
+ * @brief       AES_CMAC interface definition
+ *
+ * @author      José Ignacio Alamos <jose.alamos@inria.cl>
+ */
+
+//TODO: Add source
```
Comment (aabadie) [inline sys/include/hashes/cmac.h]: What do you mean ?

File: sys/hashes/cmac.c [hunk 13]
```diff
@@ -0,0 +1,120 @@
+/*
+ * Copyright (C) 2016 Fundación Inria Chile
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     sys_hashes
```
Comment (aabadie) [inline sys/hashes/cmac.c]: Should it be `sys_hashes_cmac` instead ?

File: sys/include/hashes/cmac.h [hunk 14]
```diff
@@ -0,0 +1,73 @@
...
+    /** last byte in last block */
+    uint32_t M_n;
+} cmac_context_t;
+
+/**
+ * @brief Initialize CMAC message digest context
+ *
+ * @param[in] ctx     Pointer to the CMAC context to initialize
+ * @param[in] key     Key to be set
+ * @param[in] key     Size of the key
```
Comment (aabadie) [inline sys/include/hashes/cmac.h]: should be keySize (generates a Doxygen warning). Also don't use LowerCamelCase but snake_case

File: sys/hashes/cmac.c [hunk 15]
```diff
@@ -0,0 +1,120 @@
...
+{
+    uint8_t i;
+    for (i = 0; i < 15; i++)
+    {
+        y[i] = (x[i] << 1) | (x[i + 1] >> 7);
+    }
+    y[15] = x[15] << 1;
+}
+
+int cmac_init(cmac_context_t *ctx, const uint8_t *key, uint8_t keySize)
```
Comment (aabadie) [inline sys/hashes/cmac.c]: keySize is LowerCamelCase and should be snake_case

**Comment by jia200x** [comment]
I'm on it :). Sorry for the delay in this.

**Comment by aabadie** [comment]
@jia200x great ! In case you missed it, there are still some remaining static tests issues: https://riot-ci.inet.haw-hamburg.de/job/RIOT_PR/job/PR-6109/7/artifact/error_static-tests.log

**Comment by jia200x** [comment]
@aabadie 
Done ;). Squashed also.

Cheers

File: sys/include/hashes/cmac.h [hunk 16]
```diff
@@ -0,0 +1,85 @@
...
+ * @brief Finalizes the CMAC message digest
+ *
+ * @param[in] ctx     Pointer to the CMAC context
+ * @param[out] digest Result location
+ *
+ */
+void cmac_final(cmac_context_t *ctx, void *digest);
+
+/**
+ * @brief   Calculate a CMAC hash from the given data
```
Comment (aabadie) [inline sys/include/hashes/cmac.h]: @jia200x this block doesn't belong to a function: so the function is missing or the block can be removed.

**Comment by aabadie** [comment]
@jia200x almost there ! Just one minor comment remaining.

File: sys/include/hashes/cmac.h (hunk 16 above)
Comment (jia200x) [inline sys/include/hashes/cmac.h]: Missed that one. Fixed

**Comment by aabadie** [comment]
@jia200x the [static tests are still failing](https://ci.riot-labs.de/devel/RIOT-OS/RIOT/6109/f0c231d94ad1956b7446ac0cccffc0352b8cb5cf/output/static_tests.txt)
Try:
```
make doc 2>&1
```
to see what's wrong with doxygen.

**Comment by jia200x** [comment]
I'm on it. Let me check

**Comment by aabadie** [comment]
@jia200x, while you are at it, can you split the history of commit messages in 2:
* `sys/hashes: add AES-CMAC implementation`
* `tests/unittests/tests-hashes: add tests for AES-CMAC`

With the remaining doxygen issues fixed, I'll give an ACK :)

**Comment by jia200x** [comment]
@aabadie I think all issues were addressed ;)

**Comment by aabadie** [comment]
ACK on my side. @kaspar030, @BytesGalore can we merge or do you have more comments ?

File: sys/hashes/cmac.c [hunk 17]
```diff
@@ -0,0 +1,119 @@
...
+    uint8_t i;
+    for (i = 0; i < 16; i++) {
+        y[i] = x[i] ^ y[i];
+    }
+}
+
+static inline void _leftshift(uint8_t *x, uint8_t *y)
+{
+    uint8_t i;
+    for (i = 0; i < 15; i++)
```
Comment (kaspar030) [inline sys/hashes/cmac.c]: please
- put variable initializer in for (...),
- use unsigned as variable,
- put opening bracket on same line

File: sys/hashes/cmac.c [hunk 18]
```diff
@@ -0,0 +1,119 @@
...
+#include <string.h>
+
+#include "crypto/ciphers.h"
+#include "hashes/cmac.h"
+
+#define MIN(a, b) a < b ? a : b
+
+static inline void _xor128(uint8_t *x, uint8_t *y)
+{
+    uint8_t i;
```
Comment (kaspar030) [inline sys/hashes/cmac.c]: initializer into for, use unsigned

File: sys/hashes/cmac.c [hunk 19]
```diff
@@ -0,0 +1,119 @@
...
+#include <inttypes.h>
+#include <stdio.h>
+#include <string.h>
+
+#include "crypto/ciphers.h"
+#include "hashes/cmac.h"
+
+#define MIN(a, b) a < b ? a : b
+
+static inline void _xor128(uint8_t *x, uint8_t *y)
```
Comment (kaspar030) [inline sys/hashes/cmac.c]: as this is in the same compilation unit, the inline statement is probably redundant. please try if removing it changes codesize.

(same for the next function)

File: sys/hashes/cmac.c [hunk 20]
```diff
@@ -0,0 +1,119 @@
+/*
+ * Copyright (C) 2016 Fundación Inria Chile
```
Comment (kaspar030) [inline sys/hashes/cmac.c]: all copyrights are 2016, is that correct?

File: sys/hashes/cmac.c [hunk 21]
```diff
@@ -0,0 +1,119 @@
...
+    y[15] = x[15] << 1;
+}
+
+int cmac_init(cmac_context_t *ctx, const uint8_t *key, uint8_t key_size)
+{
+    if (key_size != CMAC_BLOCK_SIZE) {
+        return CIPHER_ERR_INVALID_KEY_SIZE;
+    }
+
+    memset(ctx->X, 0, CMAC_BLOCK_SIZE);
```
Comment (kaspar030) [inline sys/hashes/cmac.c]: why not memset the whole structure?

File: tests/unittests/tests-hashes/tests-hashes-cmac.c [hunk 22]
```diff
@@ -0,0 +1,82 @@
...
+
+static const uint8_t CMAC_KEY[16] = { 0x2b, 0x7e, 0x15, 0x16, 0x28, 0xae, 0xd2, 0xa6, 0xab, 0xf7, 0x15, 0x88, 0x09, 0xcf, 0x4f, 0x3c };
+
+static const uint8_t TEST_EMPTY_EXP[16] = { 0xbb, 0x1d, 0x69, 0x29, 0xe9, 0x59, 0x37, 0x28, 0x7f, 0xa3, 0x7d, 0x12, 0x9b, 0x75, 0x67, 0x46 };
+
+static const uint8_t TEST_1_INP[16] = { 0x6b, 0xc1, 0xbe, 0xe2, 0x2e, 0x40, 0x9f, 0x96, 0xe9, 0x3d, 0x7e, 0x11, 0x73, 0x93, 0x17, 0x2a };
+
+static const uint8_t TEST_1_EXP[16] = { 0x07, 0x0a, 0x16, 0xb4, 0x6b, 0x4d, 0x41, 0x44, 0xf7, 0x9b, 0xdd, 0x9d, 0xd0, 0x4a, 0x28, 0x7c };
+
+static const uint8_t TEST_2_INP[40] = { 0x6b, 0xc1, 0xbe, 0xe2, 0x2e, 0x40, 0x9f, 0x96, 0xe9, 0x3d, 0x7e, 0x11, 0x73, 0x93, 0x17, 0x2a, 0xae, 0x2d, 0x8a, 0x57, 0x1e, 0x03, 0xac, 0x9c, 0x9e, 0xb7, 0x6f, 0xac, 0x45, 0xaf, 0x8e, 0x51, 0x30, 0xc8, 0x1c, 0x46, 0xa3, 0x5c, 0xe4, 0x11 };
```
Comment (kaspar030) [inline tests/unittests/tests-hashes/tests-hashes-cmac.c]: please format these sensibly so no line is >80 charactes.

File: sys/hashes/cmac.c (hunk 20 above)
Comment (aabadie) [inline sys/hashes/cmac.c]: yes

**Comment by aabadie** [comment]
@jia200x, I allowed myself to directly address, squash, rebase and push the requested changes from @kaspar030. Hope this is fine.

@kaspar030 is it ok for you ?

File: sys/hashes/cmac.c (hunk 19 above)
Comment (aabadie) [inline sys/hashes/cmac.c]: tested locally and it doesn't make any difference if one removes the inline.

**Comment by kaspar030** [comment]
> @kaspar030 is it ok for you ?

Sure!

**Comment by aabadie** [comment]
Murdock is green, let's go !

**Comment by jia200x** [comment]
hi @aabadie 

sure, it's ok. Sorry for the delay, I was focused in #5552.
Thank you for addressing those issues.

Cheers!

**Pull request merged**